lib/logfmt.c

```diff
@@ -131,3 +137,62 @@ EXPORTED char *logfmt_escape(const char *val)
 
     return buf_release(&buf);
 }
+
+EXPORTED void logfmt_init(struct logfmt *lf, const char *event)
+{
+    buf_reset(&lf->msg);
+    logfmt_push(lf, "event", event);
+}
+
+EXPORTED void logfmt_fini(struct logfmt *lf)
+{
+    buf_free(&lf->msg);
+}
+
+EXPORTED const char *logfmt_cstring(const struct logfmt *lf)
+{
+    return buf_cstring(&lf->msg);
+}
+
+EXPORTED void logfmt_push(struct logfmt *lf,
+                          const char *key,
+                          const char *value)
+{
+    char *escaped;
+
+    assert(key && *key);
+
+    escaped = logfmt_escape(value);
+    if (buf_len(&lf->msg))
+        buf_putc(&lf->msg, ' ');
+    buf_printf(&lf->msg, "%s=%s", key, escaped);
+    free(escaped);
+}
+
+EXPORTED void logfmt_push_fmt(struct logfmt *lf, const char *key,
+                              const char *valuefmt, ...)
+{
+    struct buf formatted = BUF_INITIALIZER;
+    va_list args;
+
+    va_start(args, valuefmt);
+    buf_vprintf(&formatted, valuefmt, args);
+    va_end(args);
+
+    logfmt_push(lf, key, buf_cstring(&formatted));
+
+    buf_free(&formatted);
+}
+
+EXPORTED void logfmt_push_session(struct logfmt *lf)
+{
+    const char *traceid = trace_id();
+
+    if (session_have_id()) {
+        logfmt_push(lf, "session_id", session_id());
```

Comment:
@rjbs: I spelled this `"session_id"` because that's what you used for the logfmt login logs recently, but it's different from the `"sessionid"` spelling that we use elsewhere.  I'm assuming your use of underscore was a deliberate choice to bring Cyrus logfmt in line with other FM stuff, but if it was actually just an accident and `"sessionid"` would be correct, please lmk! :)

Reply:
Let's use `sessionid` as we did, my introduction of `_` was not intentional.

Reply:
Done